# <span style="color:brown">ARTIFICIAL NEURAL NETWORKS

# Task1- Data Exploration and Preprocessing
- Loading and exploring the dataset
- Summarizing key features
- Data preprocessing (normalization, handling missing values)

In [3]:
import pandas as pd

In [4]:
alphabets = pd.read_csv(r'file:///C:\Users\dell\AppData\Local\Temp\5bfbc6a1-f634-49a5-a8ff-a87a71eefe53_Neural%20networks.zip.e53\Neural%20networks\Alphabets_data.csv')
alphabets

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  \
0          T     2     8      3       5      1     8    13      0      6   
1          I     5    12      3       7      2    10     5      5      4   
2          D     4    11      6       8      6    10     6      2      6   
3          N     7    11      6       6      3     5     9      4      6   
4          G     2     1      3       1      1     8     6      6      6   
...      ...   ...   ...    ...     ...    ...   ...   ...    ...    ...   
19995      D     2     2      3       3      2     7     7      7      6   
19996      C     7    10      8       8      4     4     8      6      9   
19997      T     6     9      6       7      5     6    11      3      7   
19998      S     2     3      4       2      1     8     7      2      6   
19999      A     4     9      6       6      2     9     5      3      1   

       xybar  x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0          6      10       8      0       8      0       8  
1         13       3       9      2       8      4      10  
2         10       3       7      3       7      3       9  
3          4       4      10      6      10      2       8  
4          6       5       9      1       7      5      10  
...      ...     ...     ...    ...     ...    ...     ...  
19995      6       6       4      2       8      3       7  
19996     12       9      13      2       9      3       7  
19997     11       9       5      2      12      2       4  
19998     10       6       8      1       9      5       8  
19999      8       1       8      2       7      2       8  

[20000 rows x 17 columns]

In [5]:
alphabets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 17 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   letter  20000 non-null  object
 1   xbox    20000 non-null  int64 
 2   ybox    20000 non-null  int64 
 3   width   20000 non-null  int64 
 4   height  20000 non-null  int64 
 5   onpix   20000 non-null  int64 
 6   xbar    20000 non-null  int64 
 7   ybar    20000 non-null  int64 
 8   x2bar   20000 non-null  int64 
 9   y2bar   20000 non-null  int64 
 10  xybar   20000 non-null  int64 
 11  x2ybar  20000 non-null  int64 
 12  xy2bar  20000 non-null  int64 
 13  xedge   20000 non-null  int64 
 14  xedgey  20000 non-null  int64 
 15  yedge   20000 non-null  int64 
 16  yedgex  20000 non-null  int64 
dtypes: int64(16), object(1)
memory usage: 2.6+ MB


- As we can see here there are 16 ints and 1 object

In [7]:
# Display first 5 records of the data set
alphabets.head()

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  xybar  \
0      T     2     8      3       5      1     8    13      0      6      6   
1      I     5    12      3       7      2    10     5      5      4     13   
2      D     4    11      6       8      6    10     6      2      6     10   
3      N     7    11      6       6      3     5     9      4      6      4   
4      G     2     1      3       1      1     8     6      6      6      6   

   x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
0      10       8      0       8      0       8  
1       3       9      2       8      4      10  
2       3       7      3       7      3       9  
3       4      10      6      10      2       8  
4       5       9      1       7      5      10

In [8]:
# Display last 5 records of the data set
alphabets.tail()

letter  xbox  ybox  width  height  onpix  xbar  ybar  x2bar  y2bar  \
19995      D     2     2      3       3      2     7     7      7      6   
19996      C     7    10      8       8      4     4     8      6      9   
19997      T     6     9      6       7      5     6    11      3      7   
19998      S     2     3      4       2      1     8     7      2      6   
19999      A     4     9      6       6      2     9     5      3      1   

       xybar  x2ybar  xy2bar  xedge  xedgey  yedge  yedgex  
19995      6       6       4      2       8      3       7  
19996     12       9      13      2       9      3       7  
19997     11       9       5      2      12      2       4  
19998     10       6       8      1       9      5       8  
19999      8       1       8      2       7      2       8

In [9]:
# Finding the missing values of dataset
alphabets.isnull().sum()

letter    0
xbox      0
ybox      0
width     0
height    0
onpix     0
xbar      0
ybar      0
x2bar     0
y2bar     0
xybar     0
x2ybar    0
xy2bar    0
xedge     0
xedgey    0
yedge     0
yedgex    0
dtype: int64

- As we can see here there are no missing values in data set

In [11]:
alphabets.columns

Index(['letter', 'xbox', 'ybox', 'width', 'height', 'onpix', 'xbar', 'ybar',
       'x2bar', 'y2bar', 'xybar', 'x2ybar', 'xy2bar', 'xedge', 'xedgey',
       'yedge', 'yedgex'],
      dtype='object')

In [12]:
# Normalizing the data
from sklearn.preprocessing import StandardScaler

features = alphabets.drop(columns=['letter']) 
scaler = StandardScaler()
normalized_features = scaler.fit_transform(features)

In [13]:
# Adding the target column back to the normalized features
processed_data = pd.DataFrame(normalized_features, columns=features.columns)
processed_data['letter'] = alphabets['letter']  # Replace 'target' with the actual target column name

# Saving the processed data to a new CSV file
processed_data.to_csv("Processed_Alphabets_data.csv", index=False)

### Summarize its key features such as the number of samples, features, and classes.

In [15]:
# Number of samples (rows) and features (columns)
num_samples = alphabets.shape[0]
num_features = alphabets.shape[1] - 1  # excluding the target column

print("\nNumber of samples:", num_samples)
print("Number of features:", num_features)


Number of samples: 20000
Number of features: 16


In [16]:
# Identify the unique classes in the target column
unique_classes = alphabets['letter'].unique()
num_classes = len(unique_classes)

print("\nNumber of unique classes:", num_classes)
print("Classes:", unique_classes)


Number of unique classes: 26
Classes: ['T' 'I' 'D' 'N' 'G' 'S' 'B' 'A' 'J' 'M' 'X' 'O' 'R' 'F' 'C' 'H' 'W' 'L'
 'P' 'E' 'V' 'Y' 'Q' 'U' 'K' 'Z']


In [17]:
# Summary statistics
print("\nSummary statistics of the dataset:")
print(alphabets.describe(include='all'))


Summary statistics of the dataset:
       letter          xbox          ybox         width       height  \
count   20000  20000.000000  20000.000000  20000.000000  20000.00000   
unique     26           NaN           NaN           NaN          NaN   
top         U           NaN           NaN           NaN          NaN   
freq      813           NaN           NaN           NaN          NaN   
mean      NaN      4.023550      7.035500      5.121850      5.37245   
std       NaN      1.913212      3.304555      2.014573      2.26139   
min       NaN      0.000000      0.000000      0.000000      0.00000   
25%       NaN      3.000000      5.000000      4.000000      4.00000   
50%       NaN      4.000000      7.000000      5.000000      6.00000   
75%       NaN      5.000000      9.000000      6.000000      7.00000   
max       NaN     15.000000     15.000000     15.000000     15.00000   

               onpix          xbar          ybar         x2bar         y2bar  \
count   20000.00000

# Task 2: Model Implementation
- Construct a basic ANN model
- Divide the dataset into training and test sets
- Train the model and make predictions

In [19]:
pip install keras

Note: you may need to restart the kernel to use updated packages.


In [20]:
pip install tensorflow

Note: you may need to restart the kernel to use updated packages.


In [21]:
import pandas as pd
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical

# Loading the new processed dataset, which we made
data = pd.read_csv("Processed_Alphabets_data.csv")

# Spliting the data into features and target
x = data.drop(columns=['letter'])
y = to_categorical(data['letter'])

ValueError: invalid literal for int() with base 10: 'T'

In [22]:
#Target var is in obj so we are getting this error.  Let's convert it into int using LabelEncoder

In [24]:
import pandas as pd
from sklearn.model_selection import train_test_split
from keras.models import Sequential
from keras.layers import Dense
from keras.utils import to_categorical
from sklearn.preprocessing import LabelEncoder

# Loading the new processed dataset, which we made
data = pd.read_csv("Processed_Alphabets_data.csv")

# Splitting the data into features and target
x = data.drop(columns=['letter'])

# Encoding string labels to integers
le = LabelEncoder()
y_int = le.fit_transform(data['letter'])

# Converting integer labels to categorical format
y = to_categorical(y_int)

In [26]:
# Split the dataset into training and test sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=42)

In [28]:
# ANN model
model = Sequential()
model.add(Dense(64, input_dim=x_train.shape[1], activation='relu'))# Hidden layer 1
model.add(Dense(32, activation='relu'))# Hidden layer 2
model.add(Dense(y.shape[1], activation='softmax'))  # Output layer with softmax activation for classification

C:\Users\dell\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


In [30]:
# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics = ['accuracy'])

In [32]:
# Train the model

model.fit(x_train, y_train, epochs=5)

Epoch 1/5
500/500 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.3279 - loss: 2.4698
Epoch 2/5
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 1ms/step - accuracy: 0.7353 - loss: 0.9266
Epoch 3/5
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.7918 - loss: 0.7158
Epoch 4/5
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8296 - loss: 0.5857
Epoch 5/5
500/500 ━━━━━━━━━━━━━━━━━━━━ 1s 2ms/step - accuracy: 0.8591 - loss: 0.4983


In [34]:
# Evaluate the model on the test set
dloss, daccuracy = model.evaluate(x_test, y_test)
print("Test accuracy =", daccuracy)

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.8618 - loss: 0.4711
Test accuracy = 0.8615000247955322


# Task 3: Hyperparameter Tuning
- Modify hyperparameters
- Use grid search or random search for tuning

In [36]:
# We need to wrap our Keras model in a KerasClassifier and then define a function to build model. 
# Since we have larger dataset, we'll use Random Seacrh instead of Grid Search

In [38]:
print(x_train.shape)
print(x_test.shape)

(16000, 16)
(4000, 16)


In [40]:
pip install keras-tuner

Note: you may need to restart the kernel to use updated packages.


In [42]:
# import numpy as np
# from tensorflow.keras.models import Sequential
# from tensorflow.keras.layers import Dense
# from tensorflow.keras.optimizers import Adam, SGD
# import keras_tuner as kt
# from sklearn.model_selection import train_test_split

# # Define a function to build the model with hyperparameters
# def build_model(hp):
#     model = Sequential()
#     model.add(Dense(units=hp.Int('units', min_value=32, max_value=128, step=32), activation='relu', input_dim=10))
#     model.add(Dense(1, activation='sigmoid'))
    
#     model.compile(
#         optimizer=hp.Choice('optimizer', values=['adam', 'sgd']),
#         loss='binary_crossentropy',
#         metrics=['accuracy']
#     )
#     return model

# # Create a tuner object
# tuner = kt.Hyperband(
#     build_model,
#     objective='val_accuracy',
#     max_epochs=10,
#     factor=3,
#     directory='my_dir',
#     project_name='intro_to_kt'
# )

# # Load and preprocess data
# # Example dummy data
# X = np.random.rand(100, 10)
# y = np.random.randint(2, size=100)

# # Split data into training and validation sets
# X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2)

# # Search for the best hyperparameters
# tuner.search(X_train, y_train, epochs=10, validation_data=(X_val, y_val))

# # Get the best hyperparameters and build the model
# best_hps = tuner.get_best_hyperparameters(num_trials=1)[0]
# print(f"Best hyperparameters: {best_hps.values}")

# # Build the model with the best hyperparameters
# model = tuner.hypermodel.build(best_hps)
# model.summary()

In [44]:
pip install tensorflow

Note: you may need to restart the kernel to use updated packages.


In [46]:
pip install scikeras

Note: you may need to restart the kernel to use updated packages.


In [48]:
import pandas as pd
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from scikeras.wrappers import KerasClassifier
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam, SGD
from sklearn.preprocessing import LabelEncoder
from tensorflow.keras.utils import to_categorical
import numpy as np

In [50]:
# Define a function to build the model
def build_model(optimizer='adam', neurons=32, activation='relu', hidden_layers=1, learning_rate=0.01):
    model = Sequential()
    model.add(Dense(neurons, input_dim=16, activation=activation))  # Adjust input_dim as per your data
    
    for _ in range(hidden_layers):
        model.add(Dense(neurons, activation=activation))
    
    model.add(Dense(y.shape[1], activation='softmax'))

    
    if optimizer == 'adam':
        optimizer = Adam(learning_rate=learning_rate)
    elif optimizer == 'SGD':
        optimizer = SGD(learning_rate=learning_rate)
    
    model.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])  # Adjust for your task
    return model

# Wrap the model using KerasClassifier
model = KerasClassifier(build_fn=build_model, verbose=0)

# Define the parameter grid
search_dict = {
    'optimizer': ['adam', 'SGD'],
    'neurons': [32, 64, 128],
    'activation': ['relu', 'tanh'],
    'batch_size': [10, 20],
    'epochs': [10, 50],
    'hidden_layers': [1, 2, 3],
    'learning_rate': [0.001, 0.01, 0.1]
}

# Perform random search
random_search = RandomizedSearchCV(estimator=model, param_distributions=search_dict, n_iter=10, cv=3, random_state=42, n_jobs=-1)
random_search_result = random_search.fit(x_train, y_train)

# Display the best parameters
print("Best parameters are:", random_search_result.best_params_)

ValueError: Invalid parameter neurons for estimator KerasClassifier.
This issue can likely be resolved by setting this parameter in the KerasClassifier constructor:
`KerasClassifier(neurons=128)`
Check the list of available parameters with `estimator.get_params().keys()`

In [54]:
import pandas as pd
from sklearn.model_selection import train_test_split, RandomizedSearchCV
from scikeras.wrappers import KerasClassifier
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import Adam, SGD
from sklearn.preprocessing import LabelEncoder
from tensorflow.keras.utils import to_categorical
import numpy as np

# Load the processed dataset
data = pd.read_csv("Processed_Alphabets_data.csv")

# Split the data into features and target
X = data.drop(columns=['letter'])

# Encode string labels to integers
label_encoder = LabelEncoder()
y_int = label_encoder.fit_transform(data['letter'])

# Convert integer labels to categorical format
y = to_categorical(y_int)

# Split the dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Define a function to build the model
def build_model(optimizer='adam', neurons=32, activation='relu', hidden_layers=1, learning_rate=0.01):
    model = Sequential()
    model.add(Dense(neurons, input_dim=X_train.shape[1], activation=activation))
    
    for _ in range(hidden_layers):
        model.add(Dense(neurons, activation=activation))
    
    model.add(Dense(y.shape[1], activation='softmax'))
    
    if optimizer == 'adam':
        optimizer = Adam(learning_rate=learning_rate)
    elif optimizer == 'SGD':
        optimizer = SGD(learning_rate=learning_rate)
    
    model.compile(optimizer=optimizer, loss='categorical_crossentropy', metrics=['accuracy'])
    return model

# Wrap the model using KerasClassifier
model = KerasClassifier(
    model=build_model,
    optimizer='adam', 
    neurons=32, 
    activation='relu', 
    hidden_layers=1, 
    learning_rate=0.01,
    batch_size=10, 
    epochs=10, 
    verbose=0
)

# Define the parameter grid
search_dict = {
    'optimizer': ['adam', 'SGD'],
    'neurons': [32, 64, 128],
    'activation': ['relu', 'tanh'],
    'batch_size': [10, 20],
    'epochs': [10, 50],
    'hidden_layers': [1, 2, 3],
    'learning_rate': [0.001, 0.01, 0.1]
}

# Perform random search
random_search = RandomizedSearchCV(estimator=model, param_distributions=search_dict, n_iter=10, cv=3, random_state=42, n_jobs=-1)
random_search_result = random_search.fit(X_train, y_train)

# Display the best parameters
print("Best parameters are:", random_search_result.best_params_)

# Evaluate the best model on the test set
best_model = random_search_result.best_estimator_
test_accuracy = best_model.score(X_test, y_test)
print(f"Test accuracy of the best model: {test_accuracy:.2f}")

C:\Users\dell\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Best parameters are: {'optimizer': 'SGD', 'neurons': 128, 'learning_rate': 0.001, 'hidden_layers': 1, 'epochs': 50, 'batch_size': 10, 'activation': 'tanh'}
Test accuracy of the best model: 0.96


## Re-building the model using the Best parameters

In [57]:
# Define the model with 3 hidden layers
model = Sequential()

# Input layer
model.add(Dense(128, input_dim=x_train.shape[1], activation='tanh'))

# Hidden layers
model.add(Dense(128, activation='tanh'))

# Output layer
model.add(Dense(y.shape[1], activation='softmax'))

# Compile the model
optimizer = Adam(learning_rate=0.001)
model.compile(optimizer=optimizer, loss='binary_crossentropy', metrics=['accuracy'])  # For binary classification

# Train the model
history = model.fit(x_train, y_train, epochs=50, batch_size=10)

Epoch 1/50


C:\Users\dell\anaconda3\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


1600/1600 ━━━━━━━━━━━━━━━━━━━━ 4s 2ms/step - accuracy: 0.5426 - loss: 0.1934
Epoch 2/50
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.7992 - loss: 0.0521
Epoch 3/50
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.8557 - loss: 0.0369
Epoch 4/50
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 2s 1ms/step - accuracy: 0.8901 - loss: 0.0297
Epoch 5/50
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.9093 - loss: 0.0245
Epoch 6/50
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9283 - loss: 0.0207
Epoch 7/50
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9410 - loss: 0.0178
Epoch 8/50
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9476 - loss: 0.0158
Epoch 9/50
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9547 - loss: 0.0142
Epoch 10/50
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9621 - loss: 0.0126
Epoch 11/50
1600/1600 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step - accuracy: 0.9644 - loss: 0.0112
Epoch 12/50
1600/1600 ━━━━━━━━━━━━━━━━━━━

In [59]:
# Evaluate the model on the test set
new_loss, new_accuracy = model.evaluate(x_test, y_test)
print("Test accuracy =", new_accuracy)

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 1ms/step - accuracy: 0.9720 - loss: 0.0099
Test accuracy = 0.9727500081062317


# Task 4: Evaluation
- Employ suitable metrics such as accuracy, precision, recall, and F1-score to evaluate your model's performance.
- Discuss the performance differences between the model with default hyperparameters and the tuned model, emphasizing the effects of hyperparameter tuning.

In [63]:
from sklearn.metrics import classification_report

# Make predictions
y_pred = model.predict(x_test)
y_pred = (y_pred > 0.5).astype(int)  # Converting probabilities to binary predictions

# Print classification report
print(classification_report(y_test, y_pred))

125/125 ━━━━━━━━━━━━━━━━━━━━ 0s 902us/step
              precision    recall  f1-score   support

           0       0.98      1.00      0.99       149
           1       0.93      0.97      0.95       153
           2       0.98      0.95      0.97       137
           3       0.96      0.98      0.97       156
           4       0.94      0.99      0.97       141
           5       0.94      0.97      0.96       140
           6       0.96      0.96      0.96       160
           7       0.97      0.92      0.95       144
           8       0.99      0.95      0.97       146
           9       0.97      0.99      0.98       149
          10       0.96      0.94      0.95       130
          11       0.99      0.98      0.98       155
          12       0.98      0.98      0.98       168
          13       0.98      0.96      0.97       151
          14       0.97      0.98      0.97       145
          15       0.99      0.96      0.97       173
          16       0.99      0.99     

C:\Users\dell\anaconda3\Lib\site-packages\sklearn\metrics\_classification.py:1509: UndefinedMetricWarning: Precision is ill-defined and being set to 0.0 in samples with no predicted labels. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, f"{metric.capitalize()} is", len(result))


# Output Explanation :

1. Precision: The proportion of positive identifications that were actually correct.

2. Recall: The proportion of actual positives that were correctly identified.

3. F1-Score: The harmonic mean of precision and recall, providing a single metric to evaluate the model’s performance.

4. Support: The number of true instances for each class

5. Micro Average: Calculates metrics globally by counting the total true positives, false negatives, and false positives. It is useful when you want to    see the overall performance of the model across all classes.

6. Macro Average: Calculates metrics for each class independently and then takes the average. It treats all classes equally, regardless of their           support.

7. Weighted Average: Calculates metrics for each class and then takes the average, weighted by the number of true instances for each class. It accounts    for class imbalance.

8. Samples Average: This is used for multi-label classification problems. It calculates metrics for each label and then averages them.

# Observations:

### High Precision, Recall, and F1-Scores: 
- It indicates that the model seems to be performing well.

### micro avg, macro avg, and weighted avg values:
- They are close, indicating balanced performance across classes and suggesting that the model is generally effective.

### Sample Size: 
- The support values show the number of samples for each class, which can help understand if certain classes are underrepresented.

    - Overall, this report indicates that the tuned model is performing well across multiple metrics and classes. :-)

In [67]:
print("--Performance Comparison--")

print("Model with Default Hyperparameters(Test Accuracy) = ", daccuracy)
print("Model with Default Hyperparameters(Loss)          = ", dloss)
print("------------------------------------------------------------------------------")
print("Model with Tuned Hyperparameters(Test Accuracy) = ", new_accuracy)
print("Model with Tuned Hyperparameters(Loss)          = ", new_loss)

--Performance Comparison--
Model with Default Hyperparameters(Test Accuracy) =  0.8615000247955322
Model with Default Hyperparameters(Loss)          =  0.48098018765449524
------------------------------------------------------------------------------
Model with Tuned Hyperparameters(Test Accuracy) =  0.9727500081062317
Model with Tuned Hyperparameters(Loss)          =  0.009098337963223457


# Analysis of Hyperparameter Tuning Effects:

## Accuracy Improvement:
Default Model: 86%

Tuned Model: 97%

Conclusion: The accuracy increased by approximately 11% after tuning. This suggests that the tuned model performs significantly better on unseen data, indicating that the tuning process helped in optimizing the model to generalize better.


## Loss Reduction:

Default Model: 0.4809

Tuned Model: 0.0090

Conclusion: The loss decreased drastically by 0.4719. A lower loss value indicates that the tuned model makes fewer errors and performs better in predicting the correct labels.

# Last Task : Evaluation Criteria
●	Accuracy and completeness of the implementation.

●	Proficiency in data preprocessing and model development.

●	Systematic approach and thoroughness in hyperparameter tuning.

●	Depth of evaluation and discussion.

●	Overall quality of the report.

# Evaluation of the ANN Model Development and Tuning Project

## 1. Accuracy and Completeness of the Implementation
### Accuracy:
- The ANN model has achieved an impressive accuracy of over 97% on the test set after performing HyperParameter Tuning. - This high accuracy indicates that the model is effectively distinguishing between different classes in the dataset. 
- It also includes key preprocessing steps, such as normalizing features and encoding categorical variables.

### Completeness:
- All necessary preprocessing steps are applied to prepare the dataset for training. The dataset was properly split into training and test sets, and features were normalized. These steps are crucial for the model's performance and have been executed thoroughly. All the errors are handled adn solved.



## 2. Proficiency in Data Preprocessing and Model Development
### Data Preprocessing(Data preprocessing is the process of cleaning and transforming raw data into a format suitable for analysis)
### Steps:
1. Firstly we have checked for any missing values. In our dataset we didn't had any missing values.
2. After that we checked for any categorical data in the dataset using info(). Through this we came to know that our dataset contains 1 object i.e., 'letter'. So we dropped that column, then we used LabelEncoder to convert that object into int, and then we stored it in a varibale names as 'y'. (as shown in Task 2)
3. We then spil the dataset into training and test sets using train_test_split(). We have given 80% to Train and remaining 20% to test

### Model Development:
- An ANN model was built with two hidden layers containing 32 and 64 neurons, respectively, and the 'relu' activation function.
- Initially we took epoch as 5 which gave a kind of good accuracy of . 
- To increase the model accuracy we did HyperParameter Tuning and based on the given result we re-run the model and got an accuracy of .


## 3. Systematic Approach and Thoroughness in Hyperparameter Tuning
### Systematic Approach:
- Since we hade a larger datset, we choose RandomizedSearchCV instead of GridSearchCV, as RandomizedSearchCV works well and ffaster for larger datset.
- We also used cross-validation(cv) to ensure that the chosen hyperparameters were validated across different subsets of the data. This provides a robust way to identify the optimal model configuration.

### Thoroughness in Hyperparameter Tuning:
- We have creted a dictionar(parameter grid) as search_dict and passed different hyperparameter like optimizer, neurons,
  activation, batch_size, epochs, hidden_layers and learning_rate through it, to get the BEST HP.
- After performing Hyperparameter Tuning, we got: "Best parameters are: {'optimizer': 'adam', 'neurons': 128, 'learning_rate': 0.001, 'hidden_layers': 3, 'epochs': 10, 'batch_size': 20, 'activation': 'relu'}". Using this we have re-build the model and got accuracy of .


## 4. Depth of Evaluation and Discussion
- The model's performance was evaluated using accuracy, precision, recall, and F1-score. The high accuracy and other metrics indicate strong performance, with a comprehensive assessment providing confidence in the model’s effectiveness.
- Comparison: The performance of the tuned model was compared with the baseline model. The tuned model showed improvements, demonstrating the value of hyperparameter optimization in enhancing model performance.

## 5. Overall Quality of the report 

## Analysis of Hyperparameter Tuning Effects:

### Accuracy Improvement:
- Default Model: 86%

- Tuned Model: 97%

Conclusion: The accuracy increased by approximately 11% after tuning. This suggests that the tuned model performs significantly better on unseen data, indicating that the tuning process helped in optimizing the model to generalize better.


## Loss Reduction:

- Default Model: 0.4809

- Tuned Model: 0.0090

Conclusion: The loss decreased drastically by 0.4719. A lower loss value indicates that the tuned model makes fewer errors and performs better in predicting the correct labels.

## Thank You!
- Hamza Khan